* [Home](https://github.com/mlangill/microbiome_helper/wiki/Home)
* [Requirements](https://github.com/mlangill/microbiome_helper/wiki/Requirements)
* [Brief description of scripts](https://github.com/mlangill/microbiome_helper/wiki/Brief-description-of-scripts)
* [Metagenomic Workflow](https://github.com/mlangill/microbiome_helper/wiki/Metagenomic-Workflow)
    * [Stitch reads](https://github.com/mlangill/microbiome_helper/wiki/Stitch-reads)
    * [Sequence QC](https://github.com/mlangill/microbiome_helper/wiki/Sequence-QC)
    * [Screen out human sequences](https://github.com/mlangill/microbiome_helper/wiki/Screen-out-human-sequences)
    * [Taxonomic composition](https://github.com/mlangill/microbiome_helper/wiki/Taxonomic-composition)
    * [Functional profiling](https://github.com/mlangill/microbiome_helper/wiki/Functional-profiling)

* [16S Workflow](https://github.com/mlangill/microbiome_helper/wiki/16S-Workflow)
    * [16S tutorial](https://github.com/mlangill/microbiome_helper/wiki/16S-tutorial)
    * [Stitch reads](https://github.com/mlangill/microbiome_helper/wiki/Stitch-reads)
    * [Sequence QC](https://github.com/mlangill/microbiome_helper/wiki/Sequence-QC)
    * [Remove low quality reads](https://github.com/mlangill/microbiome_helper/wiki/Remove-low-quality-reads)
    * [Remove chimeric reads](https://github.com/mlangill/microbiome_helper/wiki/Remove-chimeric-reads)
    * [OTU picking](https://github.com/mlangill/microbiome_helper/wiki/OTU-picking)
    * [Alpha and beta diversity](https://github.com/mlangill/microbiome_helper/wiki/Alpha-and-beta-diversity)

* [Changes to 16S workflow for 18S data](https://github.com/mlangill/microbiome_helper/wiki/Changes-to-16S-workflow-for-18S-data)

* [Additional QIIME analysis](https://github.com/mlangill/microbiome_helper/wiki/Additional-QIIME-analysis)

* [PICRUSt workflow (for 16S data)](https://github.com/mlangill/microbiome_helper/wiki/PICRUSt-workflow)