Microbiome Helper is a repository that contains several resources to help researchers working with microbial sequencing data: 
* A series of scripts that help process and automate various microbiome and metagenomic bioinformatic tools. 
* Workflows for analyzing 16S/18S rRNA and metagenomic data. 
* Tutorials with test data, example output, and questions for different microbiome analyses.
* A Virtual Box image that can be used to run our workflows and tutorials with little or no configuration.

These scripts were produced by the [Integrated Microbiome Resource](http://cgeb-imr.ca/index.html). It is important that you cite the [tools](https://github.com/mlangill/microbiome_helper/wiki/Requirements) that are wrapped by our scripts.

**Note that the scripts and workflows are continually being updated.**

You can use the sidebar menu to navigate the wiki.

### Before getting started:

* These workflows usually start with raw paired-end Illumina data in demultiplexed fastq format.

* Most of the commands allow parallel threads to be used with the -p option. Adjust according to your computer hardware abilities.
