Microbiome Helper is a repository that contains several resources to help researchers working with microbial sequencing data: 
* A series of scripts that help process and automate various microbiome and metagenomic bioinformatic tools. 
* Workflows or standard operating procedures (SOPs) for analyzing 16S/18S rRNA and metagenomic data. 
* Tutorials with test data, example output, and questions for different microbiome analyses.
* A Virtual Box image that can be used to run our workflows and tutorials with little or no configuration.
   
You can use the sidebar menu to navigate the wiki.   
  
**Note that the scripts and workflows are continually being updated.**  
  
These scripts were produced by the [Integrated Microbiome Resource](http://cgeb-imr.ca/index.html). It is important that you [cite](https://github.com/mlangill/microbiome_helper/wiki/Requirements) the tools that are wrapped by our [scripts](https://github.com/mlangill/microbiome_helper). Also, our paper on Microbiome Helper has been [published in mSystems](http://msystems.asm.org/content/2/1/e00127-16). Please cite this paper if you use our repository:  

_Comeau AM, Douglas GM, Langille MGI. 2017. Microbiome Helper: a Custom and Streamlined Workflow for Microbiome Research. mSystems 2(1): e00127-16; DOI: 10.1128/mSystems.00127-16_  
      
   
   

  

