These are the answers to the questions from the [Metagenomics Summer School Tutorial](https://github.com/mlangill/microbiome_helper/wiki/Metagenomics-Tutorial-Summer-School).  

1) 12 - since these are paired-end reads there is a forward and reverse FASTQ for each sample.
2) 6 - from the mapping file there are 6 _Cancer_ and 6 _Normal_ samples.
3) Each FASTQ is 300,000 lines long, which means there are 75,000 paired-end reads in each one.
4) p144C
5) humann2 v0.11.1 and MetaPhlAn version 2.5.0 (28 April 2015) were installed when this tutorial was written. Keeping track of this information (including for the other programs in your pipeline) is important for reproducibility.
6) 17.2%
7) 13.3%
8) g__Streptococcus
9) You should have saved a plot like this:
[[images/mgs_tutorial_2017_images/g__Streptococcus_boxplot.png]]
10) P=0.004