* [Introduction](#introduction)  
    * [Requirements](#requirements)  
    * [General Background](#background)  
    * [Other Resources](#other-resources)  
* [Important Considerations](#important-considerations)  
* [Pre-processing](#pre-processing)  
    * [Removing Rare Features](#removing-rare-features)  
    * [Transforming Your Data](#transforming-your-data)   
* [Assessing Model Fit](#assessing-model-fit)  
    * [Permutation Test](#permutation-test)  
    * [Receiver Operating Characteristic Curve](#receiver-operating-characteristic-curve)   
* [Identifying Important Features](#identifying-important-features)  
  
## Introduction
This tutorial is intended to teach beginners the basics of running random forest (RF) models on microbial sequencing data. Importantly, the below commands are **not the optimal settings to use for all datasets**. An important part of data analysis is exploring your data and deciding on what approaches would be best for your particular problem. However, as shall be described below, parameterizing models based upon your training and/or test set can resulting in biased models.  
  
Typically RF models are run when researchers are interested in classifying samples into two groups of interest based upon 
  
### Requirements
  
### Background
  
### Other Resources
* [MetAML](http://segatalab.cibio.unitn.it/tools/metaml/) - A Python packaged produced by the Segata Lab from Trento University for running machine learning on metagenomics datasets. They have written a [tutorial](https://bitbucket.org/CibioCM/metaml/wiki/Home) and [research paper](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004977) for this tool. 

## Important Considerations
  
## Pre-processing
  
### Removing Rare Features
  
### Transforming Your Data
  
## Assessing Model Fit

### Permutation Test

### Receiver Operating Characteristic Curve   
  
## Identifying Important Features  
